bagua/torch_api/data_parallel/distributed.py

```diff
@@ -0,0 +1,316 @@
+                    distinct_device_types
+                )
+            )
+        self.device_type = list(distinct_device_types)[0]
+
+        self.static_graph = False
+        self.dim = dim
+        self.module = module
+        self.device = list(self.module.parameters())[0].device
+        assert broadcast_buffers is True, "Not yet supported"
+        self.broadcast_buffers = broadcast_buffers
+        assert find_unused_parameters is False, "Not yet supported"
+        self.find_unused_parameters = find_unused_parameters
+
+        self.inner = BaguaDistributedDataParallel(
+            self.module, optimizers, algorithm, to_bagua_process_group(process_group)
+        )
+
+    @property
+    def require_backward_grad_sync(self):
+        return self.inner.require_backward_grad_sync
+
+    @property
+    def parameters_to_ignore(self):
+        return self.inner.parameters_to_ignore
+
+    def forward(self, *inputs, **kwargs):
+        output = self.module(*inputs, **kwargs)
+        return output
+
+    @contextmanager
+    def no_sync(self):
+        old_require_backward_grad_sync = self.require_backward_grad_sync
+        self.inner.require_backward_grad_sync = False
+        try:
+            yield
+        finally:
+            self.inner.require_backward_grad_sync = old_require_backward_grad_sync
+
+    def switch_bagua_setting(
+        self,
+        process_group=None,
+        optimizers: List[torch.optim.Optimizer] = [],
+        algorithm: "bagua.torch_api.algorithms.Algorithm" = GradientAllReduceAlgorithm(),
+    ):
+        self.inner = BaguaDistributedDataParallel(
+            self.module,
+            optimizers=optimizers,
+            algorithm=algorithm,
+            process_group=to_bagua_process_group(process_group),
+            bagua_module_name=self.inner.bagua_module_name,
+        )
+
+    @property
+    def bagua_algorithm(self):
+        return self.inner.bagua_algorithm
+
+    @property
+    def bagua_module_name(self):
+        return self.inner.bagua_module_name
+
+    @property
+    def bagua_optimizers(self):
+        return self.inner.bagua_optimizers
+
+    @property
+    def bagua_buckets(self):
+        return self.inner.bagua_buckets
+
+
+def DistributedDataParallel(
+    module: Module,
+    device_ids: Optional[List[Union[int, torch.device]]] = None,
+    output_device: Union[int, torch.device] = None,
+    dim: int = 0,
+    broadcast_buffers: bool = True,
+    process_group: Union[None, TorchProcessGroup] = None,
+    bucket_cap_mb: int = 25,
+    find_unused_parameters: bool = False,
+    check_reduction: bool = False,
+    gradient_as_bucket_view: bool = False,
+    # The following bagua parameters
+    optimizers: List[torch.optim.Optimizer] = [],
+    algorithm: "bagua.torch_api.algorithms.Algorithm" = GradientAllReduceAlgorithm()
+) -> Union[TorchDistributedDataParallel, DistributedDataParallel_V1_9_0]:
+    r"""
+    Implements distributed data parallelism at the module level.
```

Comment:
This function provides a PyTorch DDP (put reference here) compatible interface plus several Bagua specific parameters.

With this function, you can use Bagua for existing training scripts based on PyTorch DDP. 

Example:

```
.... fill tihs
```

Example using faster algorithms in Bagua:

```
...fill this
```